pkg/gnmi/subscribe.go

```diff
@@ -16,12 +16,290 @@
+	for {
+		select {
+		case resp := <-respChan:
+			log.Infof("Send Subscribe ONCE response to client: %v", resp)
+
+			err := stream.Send(resp)
+			if err != nil {
+				errChan <- status.Error(codes.Unimplemented, err.Error())
+			}
+
+			return
+		}
+
+	}
+
+}
+
+func (s *Server) subscribePoll(stream pb.GNMI_SubscribeServer, req *pb.SubscribeRequest, errChan chan<- error) {
+
+	log.Info("Handle a Subscribe POLL Request")
+
+	respChan := make(chan *pb.SubscribeResponse)
+	go s.processSubscribe(req, respChan, errChan)
+
+	resp := <-respChan
+	log.Infof("Send Subscribe POLL response to client: %v", resp)
+
+	err := stream.Send(resp)
+	if err != nil {
+		log.Info(err)
+		errChan <- status.Error(codes.Unimplemented, err.Error())
+		return
+	}
+	log.Info("called once")
+
+	for {
+		select {
+		default:
+			req, err := stream.Recv()
+			log.Info(err)
+			if err != nil {
+				errChan <- status.Error(codes.Unimplemented, err.Error())
+				return
+			}
+
+			switch {
+			case err == io.EOF:
+				errChan <- status.Error(codes.Unimplemented, err.Error())
+				return
+			case err != nil:
+				errChan <- err
+				return
+			case req.GetSubscribe() == nil:
+				errChan <- status.Errorf(codes.InvalidArgument, "request must contain a subscription %#v", req)
+				return
+			}
+
+			if err := s.checkEncodingAndModel(req.GetSubscribe().GetEncoding(), req.GetSubscribe().UseModels); err != nil {
+				errChan <- status.Error(codes.Unimplemented, err.Error())
+				return
+			}
+
+			log.Info("called")
+			go s.processSubscribe(req, respChan, errChan)
+
+			resp := <-respChan
+			log.Infof("Send Subscribe POLL response to client: %v", resp)
+
+			err = stream.Send(resp)
+			if err != nil {
+				errChan <- status.Error(codes.Unimplemented, err.Error())
+				return
+			}
+		}
+	}
+}
+
+func (s *Server) subscribeStream(stream pb.GNMI_SubscribeServer, req *pb.SubscribeRequest, errChan chan<- error) {
+
+	log.Info("Handle a Subscribe STREAM Request")
+	respChan := make(chan *pb.SubscribeResponse)
+
+	for {
+		select {
+		default:
+			//<-s.ConfigUpdate
+			go s.processSubscribe(req, respChan, errChan)
+
+			resp := <-respChan
+			log.Infof("Send Subscribe STREAM response to client: %v", resp)
+
+			err := stream.Send(resp)
+			if err != nil {
+				errChan <- status.Error(codes.Unimplemented, err.Error())
+				return
+			}
+		}
+	}
+
+}
+
 // Subscribe method is not implemented.
 func (s *Server) Subscribe(stream pb.GNMI_SubscribeServer) error {
-	return status.Error(codes.Unimplemented, "Subscribe is not implemented.")
+
+	req, err := stream.Recv()
+
+	switch {
+	case err == io.EOF:
+		return nil
+	case err != nil:
+		return err
+	case req.GetSubscribe() == nil:
+		return status.Errorf(codes.InvalidArgument, "request must contain a subscription %#v", req)
+	}
+
+	if err := s.checkEncodingAndModel(req.GetSubscribe().GetEncoding(), req.GetSubscribe().UseModels); err != nil {
+		return status.Error(codes.Unimplemented, err.Error())
+	}
+	errChan := make(chan error)
```

Comment:
who is actually reading from this channel ?

Reply:
Andrea, 

Nothing is finalized yet and I am exploring the usage of an error channel as well to return the error from other functions to the subscribe function. However, its usage is a little bit tricky because It is not obvious to me how interacts with response channel and goroutines and which leads to a deadlock sometimes. I am exploring some possibilities to come up with the best one.